#21198 Allow TestActor/TestKit/TestProbe to create child actors

jypma opened the pull request:
This is useful in at least two scenarios:
- Unit testing actors that communicate to their parent directly
- Testing re-creating (typically persistent) actors with the same name

johanandren commented:
OK TO TEST

jypma commented:
Added ability to set supervisor strategy for children, and defaulting to `stoppingStrategy` (since that's what you want when unit testing, right?)

manonthegithub commented:
@jypma I think that is right what is needed

patriknw commented on akka-testkit/src/main/scala/akka/testkit/TestKit.scala:
```diff
@@ -688,6 +724,48 @@ trait TestKitBase {
+   * Spawns an actor as a child of this test actor, and returns the child's ActorRef.
+   * The child will stop on failure.
+   */
+  def actorOf(props: Props): ActorRef = {
```
should perhaps be named something else that explains what it does, such as `createTestActorChild`

patriknw replied:
add to `JavaTestKit` also

jypma replied:
Since this only creates children of the test actor, and since the term "test actor" is sometimes hidden (when using plain `TestProbe`), how about just `createChild` or `childActorOf`?

I personally prefer to stick closer to akka's own terminology, hence including "actor of" in the name.

jypma commented:
@patriknw renamed to `childActorOf`, added Java API, and actually wrote the missing test for testing the supervisor behaviour.

jypma commented:
Jenkins is unhappy because of
```
[info] akka-testkit: found 3 potential binary incompatibilities while checking against com.typesafe.akka:akka-testkit_2.11:2.3.14  (filtered 18)
[error]  * synthetic method childActorOf$default$3()scala.Option in trait akka.testkit.TestKitBase is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.testkit.TestKitBase.childActorOf$default$3")
[error]  * synthetic method childActorOf$default$2()scala.Option in trait akka.testkit.TestKitBase is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.testkit.TestKitBase.childActorOf$default$2")
[error]  * method childActorOf(akka.actor.Props,scala.Option,scala.Option)akka.actor.ActorRef in trait akka.testkit.TestKitBase is present only in current version
```
but adding these methods is the point of the PR :-) How to proceed?

johanandren commented:
As it is testing utilities we don't really keep binary comp for the testkit, so you can safely mark those as filtered.

jypma commented:
OK, how do I do that?

johanandren commented:
Add the filter entries from the snippet you pasted ( `ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.testkit.TestKitBase.childActorOf$default$3")`) to `project/MiMa.scala`

jypma commented:
RETEST

jypma commented:
@johanandren `ClusterShardingSpec` failing on Jenkins but working locally, what to do?

johanandren commented:
I think patrik is looking into that, so hopefully it will be fixed in master soon (#21236)

patriknw commented on akka-testkit/src/test/scala/akka/testkit/TestProbeSpec.scala:
```diff
@@ -39,6 +41,38 @@ class TestProbeSpec extends AkkaSpec with DefaultTimeout {
+      val probe = TestProbe()
+      val child = probe.childActorOf(Props(new FailingActor), supervisorStrategy = Some(SupervisorStrategy.defaultStrategy))
+
+      eventually {
```
I'd prefer our own `awaitAssert`. It probably does the same thing, but then we are using the same timeout handling as elsewhere.

patriknw commented:
LGTM, aside from possibly changing to `awaitAssert`

jypma commented:
Woo-hoo, tests passed! Please merge at your convenience.

manonthegithub commented:
Super!

drewhk commented:
Implementation looks good. I think there should be additions to the manual though, I don't expect people  to find this by code-completion/scaladocs easily.

jypma commented:
Good point. I'll try to write a little something.

patriknw commented:
can perhaps be added to the [Testing parent-child relationships](http://doc.akka.io/docs/akka/2.4/scala/testing.html#Testing_parent-child_relationships) ?

jypma commented:
@patriknw To me, that part of the manual is trying to explain two different things under one topic:
1) Testing the creation of child actors, from a parent's perspective
2) Testing an actor that is hardcoded to communicate with its parent

Would it make sense to split it into two subheadings?

patriknw commented:
sure, whatever makes it nice

jypma commented:
Wrote some docs. Quick check?

johanandren commented:
```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/java/code/docs/testkit/ParentChildTest.java:194: method childActorOf in class akka.testkit.TestKit cannot be applied to given types;
[error]   required: akka.actor.Props,scala.Option<java.lang.String>,scala.Option<akka.actor.SupervisorStrategy>
[error]   found: akka.actor.Props
[error]   reason: actual and formal argument lists differ in length
[error]     ActorRef child = parent.childActorOf(Props.create(Child.class));
```

johanandren commented on akka-testkit/src/main/scala/akka/testkit/TestKit.scala:
```diff
@@ -688,6 +725,17 @@ trait TestKitBase {
+    testActor ! Spawn(props, name, supervisorStrategy)
+    expectMsgType[ActorRef]
+  }
+
```
Do we need a Java API as well, using `Optional`?

jypma replied:
Mmm I thought I created methods with several arguments for the Java API. Might be this particular test code is using the scala `TestProbe` instead of the java one? I'll have a look.

jypma replied:
Yes it is. I'll try to rewrite using `JavaTestKit`.

johanandren replied:
You are right, I missed those but I think all variations are covered with the overloads in `JavaTestKit`

johanandren commented:
LGTM (as soon as the test code compiles and passes ofc)

jypma commented:
Quick thought: would the scala signature not be more convenient as
```
  def childActorOf(props: Props, name: String = "", supervisorStrategy: SupervisorStrategy = stoppingStrategy): ActorRef = ...
```
? That way, you don't have to add a pointless `Some(...)` when giving the default args a value.

johanandren commented:
Strategy sounds fine, but using empty string a non-value-use-default doesn't feel right to me.

jypma commented:
True, but what would be the alternative? Keep it as `Some[String]` for the name?

johanandren commented:
Or provide an overload without the name.

jypma commented:
Yeah OK, but then we'd drop the varargs completely, since scalac doesn't like varargs + overloads. Might be for the best though. I'll give it a shot.

jypma committed:
+tes #21198 Allow TestActor/TestKit/TestProbe to create child actors

This is useful in at least two scenarios:
- Unit testing actors that communicate to their parent directly
- Testing re-creating (typically persistent) actors with the same name

jypma commented:
Test passed, API touched up. Please merge :-)

johanandren commented:
Great. LGTM!

patriknw commented:
LGTM

patriknw commented:
Thanks @jypma

patriknw commented:
Refs #21198

jypma commented:
Thanks a lot for the comments and getting it in!
